flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetRecordReader.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.parquet.io.RecordReader;
+import org.apache.parquet.io.api.RecordMaterializer;
+import org.apache.parquet.io.api.RecordMaterializer.RecordMaterializationException;
+import org.apache.parquet.schema.MessageType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.CheckReturnValue;
+import javax.annotation.meta.When;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.parquet.Preconditions.checkNotNull;
+
+/**
+ * Customized {@link org.apache.parquet.hadoop.ParquetRecordReader} that support start read from particular position.
+ */
+public class ParquetRecordReader<T> {
+	private static final Logger LOG = LoggerFactory.getLogger(ParquetRecordReader.class);
+
+	private ColumnIOFactory columnIOFactory;
+	private Filter filter;
+
+	private MessageType readSchema;
+	private MessageType fileSchema;
+	private ReadSupport<T> readSupport;
+
+	private RecordMaterializer<T> recordMaterializer;
+	private T currentValue;
+	private long total;
+	private long current = 0;
+	private int currentBlock = -1;
+	private ParquetFileReader reader;
+	private RecordReader<T> recordReader;
+	private boolean strictTypeChecking = true;
+	private long totalCountLoadedSoFar = 0;
+
+	public ParquetRecordReader(ReadSupport<T> readSupport, MessageType readSchema, Filter filter) {
+		this.readSupport = readSupport;
+		this.readSchema = readSchema;
+		this.filter = checkNotNull(filter, "filter");
+	}
+
+	public ParquetRecordReader(ReadSupport<T> readSupport, MessageType readSchema) {
+		this(readSupport, readSchema, FilterCompat.NOOP);
+	}
+
+	public void initialize(ParquetFileReader reader, Configuration configuration) {
+		this.reader = reader;
+		FileMetaData parquetFileMetadata = reader.getFooter().getFileMetaData();
+		this.fileSchema = parquetFileMetadata.getSchema();
+		Map<String, String> fileMetadata = parquetFileMetadata.getKeyValueMetaData();
+		ReadSupport.ReadContext readContext = readSupport.init(new InitContext(
+			configuration, toSetMultiMap(fileMetadata), readSchema));
+
+		this.columnIOFactory = new ColumnIOFactory(parquetFileMetadata.getCreatedBy());
+		this.readSchema = readContext.getRequestedSchema();
+		this.recordMaterializer = readSupport.prepareForRead(
+			configuration, fileMetadata, readSchema, readContext);
+		this.total = reader.getRecordCount();
+		reader.setRequestedSchema(readSchema);
+	}
+
+	private void checkRead() throws IOException {
+		if (current == totalCountLoadedSoFar) {
+			PageReadStore pages = reader.readNextRowGroup();
+			recordReader = createRecordReader(pages);
+			totalCountLoadedSoFar += pages.getRowCount();
+			currentBlock++;
+		}
+	}
+
+	public void close() throws  IOException {
+		if (reader != null) {
+			reader.close();
+		}
+	}
+
+	@CheckReturnValue(when = When.NEVER)
+	public T nextRecord() {
+		return currentValue;
```

Comment:
It is mandatory that `hasNextRecord()` is called before `nextRecord()`. 
Can we change `nextRecord()` method to call `hasNextRecord()` if it wasn't called before?